### PR TITLE
feat(nvidia): Add cupti support

### PR DESCRIPTION
**Pull request by cvonelm**
This commit adds a --nvidia option, which injects a library into the program under measurement, which records entry and exit into CUDA kernels via CUPTI

We might think about bumping the CMake requirement to 3.24 with this version, as older FindCUDAToolkit.cmake fail to correctly detect CUPTI headers[1].

This implements #294

[1]  https://gitlab.kitware.com/cmake/cmake/-/merge_requests/7608

### CODE REVIEW

File: src/monitor/process_monitor_main.cpp [hunk 1]
```diff
@@ -82,14 +92,22 @@ namespace monitor
...
                    });
     tmp.push_back(nullptr);
+    std::transform(env.begin(), env.end(), std::back_inserter(c_env),
+                   [](const std::string& s)
+                   {
+                       char* pc = new char[s.size() + 1];
+                       std::strcpy(pc, s.c_str());
+                       return pc;
+                   });
+    c_env.push_back(nullptr);
```
Comment (bmario) [inline src/monitor/process_monitor_main.cpp:102]: Extract these c hacking transform things into a function.

File: src/trace/trace.cpp [hunk 2]
```diff
@@ -489,6 +503,15 @@ otf2::writer::local& Trace::create_metric_writer(const std::string& name)
     return archive_(location);
 }
 
+otf2::definition::calling_context& Trace::cuda_calling_context(std::string& file,
+                                                               std::string& function)
+{
+    LineInfo info = LineInfo::for_function(file.c_str(), function.c_str(), 0, "");
```
Comment (bmario) [inline src/trace/trace.cpp:509]: Can we get a line number from cupti?

File: include/lo2s/cupti/reader.hpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+class Reader
+{
+public:
+    Reader(trace::Trace& trace, Process process)
+    : process_(process), trace_(trace), time_converter_(perf::time::Converter::instance()),
+      ringbuf_reader_(std::to_string(process.as_pid_t()), 512)
+    {
+        struct itimerspec tspec;
+        memset(&tspec, 0, sizeof(struct itimerspec));
+        tspec.it_value.tv_nsec = 1;
```
Comment (bmario) [inline include/lo2s/cupti/reader.hpp:55]:
```suggestion
```

File: include/lo2s/cupti/reader.hpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+        tspec.it_interval.tv_sec =
+            std::chrono::duration_cast<std::chrono::seconds>(config().userspace_read_interval)
+                .count();
+
+        tspec.it_interval.tv_nsec =
+            (config().userspace_read_interval % std::chrono::seconds(1)).count();
```
Comment (bmario) [inline include/lo2s/cupti/reader.hpp:62]:
```suggestion
        tspec.it_interval.tv_nsec =
            (config().userspace_read_interval % std::chrono::seconds(1)).count();

```

File: include/lo2s/cupti/reader.hpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        timer_fd_ = timerfd_create(CLOCK_MONOTONIC, TFD_NONBLOCK);
+
+        timerfd_settime(timer_fd_, TFD_TIMER_ABSTIME, &tspec, NULL);
```
Comment (bmario) [inline include/lo2s/cupti/reader.hpp:65]: This is the third reincarnation of building a tspec and using timerfd_settime. Refactor all these into a function taking a std::chrono::duration instead.

File: include/lo2s/cupti/reader.hpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+{
+namespace cupti
+{
+
+class Reader
+{
+public:
+    Reader(trace::Trace& trace, Process process)
+    : process_(process), trace_(trace), time_converter_(perf::time::Converter::instance()),
+      ringbuf_reader_(std::to_string(process.as_pid_t()), 512)
```
Comment (bmario) [inline include/lo2s/cupti/reader.hpp:51]: just using the the pid as filename. The manpage says this:

> For portable use, a shared memory object should be identified  by  a  name  of the form /somename; that is, a null-terminated string of up to NAME_MAX (i.e., 255) characters consisting of an initial slash, followed by one or more characters, none of which are slashes.

Also, it should contain lo2s and cupti at some point.

Comment (bmario) [inline include/lo2s/cupti/reader.hpp:51]: Hard-coded size, make it a command-line option.

File: include/lo2s/cupti/reader.hpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        return timer_fd_;
+    }
+
+private:
+    Process process_;
+    trace::Trace& trace_;
+    perf::time::Converter& time_converter_;
+    RingBufReader ringbuf_reader_;
+    int timer_fd_;
+    std::string exe;
```
Comment (bmario) [inline include/lo2s/cupti/reader.hpp:109]:
```suggestion
    std::string executable_name_;
```

File: include/lo2s/cupti/reader.hpp [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            struct cupti_event_kernel* kernel = (struct cupti_event_kernel*)kernel_rb.get();
+
+            auto& writer = trace_.cuda_writer(Thread(process_.as_thread()));
+
+            std::string kernel_name = kernel->name;
+            auto& cu_cctx = trace_.cuda_calling_context(exe, kernel_name);
+
+            writer << otf2::event::calling_context_enter(time_converter_(kernel->start), cu_cctx,
+                                                         2);
+            writer << otf2::event::calling_context_leave(time_converter_(kernel->end), cu_cctx);
```
Comment (bmario) [inline include/lo2s/cupti/reader.hpp:92]: Use `writer.write_calling_context_enter/leave()` instead.

File: include/lo2s/cupti/reader.hpp [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    void read()
+    {
+        rb_ptr header_rb = ringbuf_reader_.get(sizeof(struct cupti_event_header));
+
+        if (!header_rb)
+        {
+            return;
+        }
+
+        struct cupti_event_header* header = (struct cupti_event_header*)header_rb.get();
```
Comment (bmario) [inline include/lo2s/cupti/reader.hpp:78]: evil c cast

File: include/lo2s/cupti/reader.hpp [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+        if (header->type == CuptiEventType::CUPTI_KERNEL)
+        {
+            rb_ptr kernel_rb = ringbuf_reader_.get(header->size);
+            struct cupti_event_kernel* kernel = (struct cupti_event_kernel*)kernel_rb.get();
```
Comment (bmario) [inline include/lo2s/cupti/reader.hpp:83]: evil c cast

File: include/lo2s/cupti/ringbuf.hpp [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+    uint64_t version;
+    uint64_t size;
+    std::atomic_uint64_t head;
+    std::atomic_uint64_t tail;
+    std::atomic_uint64_t fill;
+};
+
+enum class CuptiEventType : uint64_t
+{
+    CUPTI_KERNEL = 0,
```
Comment (bmario) [inline include/lo2s/cupti/ringbuf.hpp:53]: I wouldn't choose 0 as first type index but at least 1.

File: include/lo2s/cupti/ringbuf.hpp [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+};
+
+struct cupti_event_header
+{
+    CuptiEventType type;
+    uint64_t size;
+};
```
Comment (bmario) [inline include/lo2s/cupti/ringbuf.hpp:60]: I think, besides `CUPTI_KERNEL`, there's no need to smurf this with cupti here.

File: include/lo2s/cupti/ringbuf.hpp [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+        ringbuf_->fill.fetch_add(reserved_size_);
+
+        reserved_ = nullptr;
+        reserved_size_ = 0;
+
+        return true;
+    }
+
+private:
+    std::string filename_;
```
Comment (bmario) [inline include/lo2s/cupti/ringbuf.hpp:173]: unused

File: include/lo2s/cupti/ringbuf.hpp [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+public:
+    RingBufWriter(std::string filename, size_t size) : filename_(filename)
+    {
+        fd_ = shm_open(filename.c_str(), O_RDWR, 0600);
+        if (fd_ == -1)
+        {
+            throw std::system_error(errno, std::system_category());
+        }
+
+        ringbuf_ = (struct ringbuf_header*)mmap(NULL, sizeof(struct ringbuf_header) + size,
```
Comment (bmario) [inline include/lo2s/cupti/ringbuf.hpp:111]: evil c ast

File: src/cupti/lib.cpp
Comment (bmario) [inline src/cupti/lib.cpp:1]: This file need comments. I have no idea what is happening here.

**Review by bmario — CHANGES_REQUESTED** [review]

File: include/lo2s/cupti/ringbuf.hpp [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+public:
+    RingBufReader(std::string filename, size_t size)
+    {
+        fd_ = shm_open(filename.c_str(), O_CREAT | O_EXCL | O_RDWR, 0600);
+        if (fd_ == -1)
+        {
+            throw std::system_error(errno, std::system_category());
+        }
+
+        ringbuf_ = (struct ringbuf_header*)mmap(NULL, sizeof(struct ringbuf_header) + size,
```
Comment (bmario) [inline include/lo2s/cupti/ringbuf.hpp:192]: There is `EventReader::init_mmap()`, which almost does the same thing, but the right way.

**Review by bmario — CHANGES_REQUESTED** [review]
After disccussing the ringbuffer a bit with @tilsche, we think it would be better to not let individual events split up on the wrap up, thus removing the need for all the mallocs, tracking thereof and simplifying the code at the same time. Also, `fill` shouldn't be necessary.

File: include/lo2s/cupti/ringbuf.hpp (hunk 15 above)
Comment (cvonelm) [reply include/lo2s/cupti/ringbuf.hpp:192]: We either need `fill` or to leave one byte empty between head and tail, because `head == tail` could mean both full and empty. And I personally prefer fill over leaving one byte free because the latter feels like it would invite off-by-one errors

File: src/cupti/lib.cpp [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+    }
+    catch (std::exception const& e)
+    {
+        std::cerr << "Could not read lo2s ringbuffer size!";
+        exit(EXIT_FAILURE);
+    }
+
+    atexit(&atExitHandler);
+
+    cuptiSubscribe(&subscriber, (CUpti_CallbackFunc)callbackHandler, NULL);
```
[the remark this answers is not in the document]
Comment (cvonelm) [reply src/cupti/lib.cpp:185]: the call creates  the subscriber, so  we cant pass it as userData to the same cupti call, or am I missing something

File: src/monitor/process_monitor.cpp [hunk 17]
```diff
@@ -38,10 +38,11 @@ void ProcessMonitor::insert_process(Process parent, Process process, std::string
                                     bool spawn)
 {
     trace_.add_process(parent, process, proc_name);
-    insert_thread(process, process.as_thread(), proc_name, spawn);
+    insert_thread(process, process.as_thread(), proc_name, spawn, true);
 }
 
-void ProcessMonitor::insert_thread(Process process, Thread thread, std::string name, bool spawn)
+void ProcessMonitor::insert_thread(Process process, Thread thread, std::string name, bool spawn,
+                                   bool is_process)
```
[the remark this answers is not in the document]
Comment (cvonelm) [reply src/monitor/process_monitor.cpp:45]: The ScopeMonitor is created in insert_thread. And CUDA sampling should only be set up for processes not threads

File: src/trace/trace.cpp (hunk 2 above)
Comment (cvonelm) [reply src/trace/trace.cpp:509]: nope

File: include/lo2s/cupti/ringbuf.hpp (hunk 15 above)
Comment (tilsche) [reply include/lo2s/cupti/ringbuf.hpp:192]: I'd rather "waste" one byte in the buffer than have one atomic 8-byte header thing to take care of.
You take care of off-by-one errors with a proper function to count the free space.

See also https://www.kernel.org/doc/Documentation/circular-buffers.txt

P.S. I do quite frankly wonder why one would not just keep incrementing head/tail indefinitely and only `%` them upon access rather than wrap-around. That would seem to be more elegant to me - and you could use the precious last byte! But I cannot oversee the consequences and would suggest to rather go with the proven approach.

**Review by bmario — CHANGES_REQUESTED** [review]
I have these comments still pending. Don't know if they still apply.

File: include/lo2s/ringbuf.hpp [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+        header_ = first_mapping_.as<struct ringbuf_header*>();
+        start_ = first_mapping_.as<std::byte*>() + pagesize;
+
+        if (create)
+        {
+            header_->version = RINGBUF_VERSION;
+            header_->size = ringbuf_size;
+            header_->tail.store(0);
+            header_->head.store(0);
+        }
```
Comment (bmario) [inline include/lo2s/ringbuf.hpp:163]: ringbuf version is never checked.

File: include/lo2s/ringbuf.hpp [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+    uint64_t head()
+    {
+        return header_->head.load();
+    }
+
+    uint64_t tail()
+    {
+        return header_->tail.load();
+    }
+    void head(uint64_t new_head)
```
Comment (bmario) [inline include/lo2s/ringbuf.hpp:174]:
```suggestion

    void head(uint64_t new_head)
```

File: include/lo2s/ringbuf.hpp [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+    {
+        return header_->head.store(new_head);
+    }
+
+    void tail(uint64_t new_tail)
+    {
+        return header_->tail.store(new_tail);
+    }
+
+    uint64_t ringbuf_size()
```
Comment (bmario) [inline include/lo2s/ringbuf.hpp:184]:
```suggestion
    uint64_t size()
```

File: include/lo2s/ringbuf.hpp [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+
+    ~Mmap()
+    {
+        munmap(addr_, size_);
+    }
+
+private:
+    void* addr_ = nullptr;
+    size_t size_;
+};
```
Comment (bmario) [inline include/lo2s/ringbuf.hpp:106]: Extract this as a separate PR and replace `EventReader::init_mmap()` with it?

Comment (bmario) [reply include/lo2s/ringbuf.hpp:106]: At the very least, extract into separate file

**Review by bmario — APPROVED** [review]

**Review by bmario — CHANGES_REQUESTED** [review]
I blame vscode

File: include/lo2s/ringbuf.hpp (hunk 21 above)
Comment (cvonelm) [reply include/lo2s/ringbuf.hpp]: https://github.com/tud-zih-energy/lo2s/pull/341

File: include/lo2s/ringbuf.hpp [hunk 22]
```diff
@@ -0,0 +1,230 @@
...
+        if (!can_be_loaded(size))
+        {
+            return nullptr;
+        }
+        return start_ + tail();
+    }
+
+    void pop(size_t ev_size)
+    {
+        // Calling pop() without trying to get() data from the ringbuffer first is a error
```
Comment (bmario) [inline include/lo2s/ringbuf.hpp:213]:
```suggestion
        // Calling pop() without trying to get() data from the ringbuffer first is an error
```

**Review by bmario — CHANGES_REQUESTED** [review]

File: src/trace/trace.cpp [hunk 23]
```diff
@@ -411,6 +410,21 @@ otf2::writer::local& Trace::sample_writer(const ExecutionScope& writer_scope)
     return archive_(location(writer_scope));
 }
 
+otf2::writer::local& Trace::cuda_writer(const Thread& thread)
+{
+    MeasurementScope scope = MeasurementScope::cuda(thread.as_scope());
+
+    const auto& cuda_location_group = registry_.emplace<otf2::definition::location_group>(
+        ByMeasurementScope(scope), intern(scope.name()), otf2::common::location_group_type::process,
```
[the remark this answers is not in the document]
Comment (cvonelm) [reply src/trace/trace.cpp]: I have changed those to accelerator_(stream) and while it does not seem to change anything in vampir, it does seem to be more correct.

File: include/lo2s/ringbuf.hpp [hunk 24]
```diff
@@ -0,0 +1,230 @@
...
+    RingBufWriter(std::string component, pid_t pid, bool create, size_t pages = 0)
+    : ShmRingbuf(component, pid, create, pages)
+    {
+    }
+
+    std::byte* reserve(size_t ev_size)
+    {
+        // No other reservation can be active!
+        assert(reserved_size_ == 0);
+
```
Comment (bmario) [inline include/lo2s/ringbuf.hpp:168]: if ev_size == 0 return nullptr

File: include/lo2s/ringbuf.hpp [hunk 25]
```diff
@@ -0,0 +1,230 @@
...
+{
+public:
+    RingBufReader(std::string component, pid_t pid, bool create, size_t pages = 0)
+    : ShmRingbuf(component, pid, create, pages)
+    {
+    }
+
+    std::byte* get(size_t size)
+    {
+        if (!can_be_loaded(size))
```
Comment (bmario) [inline include/lo2s/ringbuf.hpp:204]: if size == 0 return nullptr

File: include/lo2s/ringbuf.hpp [hunk 26]
```diff
@@ -0,0 +1,230 @@
...
+        {
+            return nullptr;
+        }
+        return start_ + tail();
+    }
+
+    void pop(size_t ev_size)
+    {
+        // Calling pop() without trying to get() data from the ringbuffer first is an error
```
Comment (bmario) [inline include/lo2s/ringbuf.hpp:213]: if size ==0 return (or assert size > 0)

File: src/monitor/process_monitor_main.cpp [hunk 27]
```diff
@@ -117,6 +119,19 @@ static void drop_privileges()
     assert(getgid() != 0);
 }
 
+std::vector<char*> to_vector_of_c_str(std::vector<std::string> vec)
```
Comment (bmario) [inline src/monitor/process_monitor_main.cpp:122]:
```suggestion
std::vector<char*> to_vector_of_c_str(const std::vector<std::string>& vec)
```

**Review by bmario — CHANGES_REQUESTED** [review]

**Review by bmario — APPROVED** [review]
